libs/x-content/src/main/java/org/elasticsearch/xcontent/XContentBuilder.java

```diff
@@ -1302,4 +1304,41 @@ private static void ensureNoSelfReferences(final Iterable<?> value, Object origi
         }
     }
 
+    /**
+     * Checks whether the given value is writeable as x-content.
+     *
+     * If the value cannot be passed to {@link #value(Object)} without
+     * error, an error {@link IllegalArgumentException} is thrown.
+     */
+    public static void ensureToXContentable(@Nullable Object value) {
+        if (value == null) {
+            return;
+        }
+
+        if (value instanceof Object[] array) {
+            for (Object v : array) {
+                ensureToXContentable(v);
+            }
+        } else if (value instanceof Map<?,?> map) {
+            for (Map.Entry<?, ?> entry : map.entrySet()) {
+                ensureToXContentable(entry.getKey());
+                ensureToXContentable(entry.getValue());
+            }
+        } else if (value instanceof Path) {
+            // Path implements Iterable<Path> and causes endless recursion and a StackOverFlow if treated as an Iterable here
+            return;
+        } else if (value instanceof Iterable<?> iterable) {
+            // Iterable also implicitly handles Set and List
+            for (Object v : iterable) {
+                ensureToXContentable(v);
+            }
+        } else if (value instanceof ToXContent) {
+            return;
+        } else {
```

Comment:
Is it intentional that there is no `Enum` case? (`unknownValue` has one)

Reply:
Not intentional. Added.